File: python/malon_agent/malon_agent/comm/dh.py
from typing import Optional
from malon_common.protocol.base_pb2 import BaseMsg, ClientMsg, HandshakeMsg, ErrorType
from .client import Client
from malon_common.crypto.dh import KeyExchange, get_client_id
from malon_common.crypto.sym import SymmetricCipher

class DHClient:
    def __init__(self, ke: KeyExchange, client: Client):
        self._ke = ke
        self._client = client
        self._shared_key_cipher: Optional[SymmetricCipher] = None
    
    def get_client_id(self) -> bytes:
        return get_client_id(self._ke.get_public_key())

    def negotiate_key(self):
        msg = BaseMsg()
        msg.HandshakeMsg.PublicKey = self._ke.get_public_key()

        response = self._client.send_msg(msg.SerializeToString())
        response_msg = BaseMsg()
        response_msg.ParseFromString(response)

        shared_key = self._ke.get_shared_key(response_msg.HandshakeMsg.PublicKey)
        self._shared_key_cipher = SymmetricCipher(shared_key)

    def send_msg(self, msg: bytes) -> bytes:
        if self._shared_key_cipher is None:
            self.negotiate_key()
        
        client_msg = BaseMsg()
        client_msg.ClientMsg.ClientID = self.get_client_id()
        client_msg.ClientMsg.Payload = self._shared_key_cipher.encrypt_sign_msg(msg)

        response = self._client.send_msg(client_msg.SerializeToString())

        response_msg = BaseMsg()
        response_msg.ParseFromString(response)

        if response_msg.HasField('ServerMsg'):
            payload = self._shared_key_cipher.verify_decrypt_msg(response_msg.ServerMsg.Payload)
            if payload is not None:
                return payload
            else:
                raise RuntimeError('Message did not pass verification')
        elif response_msg.HasField('ErrorMsg'):
            if response_msg.ErrorMsg.type == ErrorType.HANDSHAKE_EXPIRED:
                self.negotiate_key()
                return self.send_msg(msg)
            else:
                raise RuntimeError('Server returned unknown error')
        else:
            raise RuntimeError('Unexpected message type')

File: agent/comm/dh.go
package comm

import (
	"encoding/hex"
	"encoding/json"
	"fmt"

	"../cryptoutil/dh"
	"../cryptoutil/sym"
)

type DHClient struct {
	sharedKey   []byte
	keyExchange *dh.KeyExchange
	subclient   Client
}

type HandshakeMsg struct {
	PublicKey []byte `json:"public_key"`
}

type ServerMsg struct {
	Payload []byte `json:"payload"`
}

type ClientMsg struct {
	ClientID string `json:"client_id"`
	Payload  []byte `json:"payload"`
}

type ErrorMsg struct {
	Type string `json:"type"`
}

type BaseMsg struct {
	HandshakeMsg *HandshakeMsg `json:"handshake_msg,omitempty"`
	ClientMsg    *ClientMsg    `json:"client_msg,omitempty"`
	ServerMsg    *ServerMsg    `json:"server_msg,omitempty"`
	ErrorMsg     *ErrorMsg     `json:"error_msg,omitempty"`
}

func NewDHClient(keyExchange *dh.KeyExchange, subclient Client) *DHClient {
	return &DHClient{
		sharedKey:   nil,
		keyExchange: keyExchange,
		subclient:   subclient,
	}
}

func (c *DHClient) sendBaseMsg(baseMsg *BaseMsg) (*BaseMsg, error) {
	var responseBaseMsg BaseMsg
	msg, err := json.Marshal(baseMsg)
	if err != nil {
		return nil, err
	}
	response, err := c.subclient.SendMsg(msg)
	if err != nil {
		return nil, err
	}
	err = json.Unmarshal(response, &responseBaseMsg)
	if err != nil {
		return nil, err
	}
	return &responseBaseMsg, nil
}

func (c *DHClient) NegotiateKey() error {
	msg := BaseMsg{
		HandshakeMsg: &HandshakeMsg{
			PublicKey: c.keyExchange.GetPublicKey(),
		},
	}
	response, err := c.sendBaseMsg(&msg)
	if err != nil {
		return err
	}

	if handshakeMsg := response.HandshakeMsg; handshakeMsg != nil {
		c.sharedKey, err = c.keyExchange.GetSharedKey(handshakeMsg.PublicKey)
		if err != nil {
			return err
		}
	} else {
		return fmt.Errorf("expected HandshakeMsg")
	}

	return nil
}

func (c *DHClient) GetClientID() []byte {
	return dh.GetClientID(c.keyExchange.GetPublicKey())
}

func (c *DHClient) SendMsg(msg []byte) ([]byte, error) {
	if c.sharedKey == nil {
		err := c.NegotiateKey()
		if err != nil {
			return nil, err
		}
	}

	encryptedMsg, err := sym.EncryptThenSignMessage(msg, c.sharedKey)
	if err != nil {
		return nil, err
	}

	baseMsg := &BaseMsg{
		ClientMsg: &ClientMsg{
			ClientID: hex.EncodeToString(c.GetClientID()),
			Payload:  encryptedMsg,
		},
	}
	responseBaseMsg, err := c.sendBaseMsg(baseMsg)
	if err != nil {
		return nil, err
	}

	if serverMsg := responseBaseMsg.ServerMsg; serverMsg != nil {
		return sym.ValidateThenDecryptMessage(serverMsg.Payload, c.sharedKey)
	} else if errorMsg := responseBaseMsg.ErrorMsg; errorMsg != nil {
		if errorMsg.Type == "HANDSHAKE_EXPIRED" {
			c.NegotiateKey()
			return c.SendMsg(msg)
		} else {
			return nil, fmt.Errorf("server returned unknown error")
		}
	} else {
		return nil, fmt.Errorf("expected ServerMsg")
	}
}
